qiita_plugins/target_gene/tgp/qiita_client.py

```diff
@@ -0,0 +1,197 @@
+        self._fetch_token()
+
+    def _fetch_token(self):
+        """Retrieves an access token from the Qiita server
+
+        Raises
+        ------
+        ValueError
+            If the authentication with the Qiita server fails
+        """
+        data = {'client_id': self._client_id,
+                'client_secret': self._client_secret,
+                'grant_type': 'client'}
+        r = requests.post(self._authenticate_url, verify=self._verify,
+                          data=data)
+        if r.status_code != 200:
+            raise ValueError("Can't authenticate with the Qiita server")
+        self._token = r.json()['access_token']
+
+    def _request_oauth2(self, req, *args, **kwargs):
+        """Executes a request using OAuth2 authorization
+
+        Parameters
+        ----------
+        req : function
+            The request to execute
+        args : tuple
+            The request args
+        kwargs : dict
+            The request kwargs
+
+        Returns
+        -------
+        requests.Response
+            The request response
+        """
+        if 'headers' in kwargs:
+            kwargs['headers']['Authorization'] = 'Bearer %s' % self._token
+        else:
+            kwargs['headers'] = {'Authorization': 'Bearer %s' % self._token}
+        r = req(*args, **kwargs)
+        r.close()
+        if r.status_code == 400:
+            if r.json()['error_description'] == \
+                    'Oauth2 error: token has timed out':
+                # The token expired - get a new one and re-try the request
+                self._fetch_token()
+                r = req(*args, **kwargs)
+        return r
+
+    def _request_retry(self, req, url, **kwargs):
+        """Executes a request retrying it 2 times in case of failure
+
+        Parameters
+        ----------
+        req : function
+            The request to execute
+        url : str
+            The url to access in the server
+        kwargs : dict
+            The request kwargs
+
+        Returns
+        -------
+        dict
+            The JSON information in the request response
+
+        Notes
+        -----
+        After doing some research on the topic, there are multiple ways of
+        engineering the number of times a request should be retried (multiple
+        sources - most of them on RPC systems). A short summary of those are:
+          1. Keep retrying indefinitely
+          2. The probability of retrying a request is based on the number of
+          retries already done, as well as the cost of a retry
+          3. Retry just once
+
+        Number 1 could create an infinite loop. Number 2 is too complex and
+        the cost of retrying depends on the actual work that we are currently
+        doing (which is unknown to the current function). We thus decided to
+        implement 3, which is simple and allows to overcome simple
+        communication problems.
+        """
+        url = self._server_url + url
```

Comment:
I disagree as those are not paths and, although we don't support it, on windows it will not work.